src/System.IO.Pipelines/BufferSegment.cs

```diff
@@ -85,11 +85,11 @@ public BufferSegment(OwnedMemory<byte> buffer, int start, int end)
 
         public void Dispose()
         {
-            Debug.Assert(_buffer.ReferenceCount >= 1);
+            Debug.Assert(!_buffer.HasZeroReferences);
```

Comment:
`AutoReference` sounds like it deals with disposal for you; would need something like "Destructible Types" https://github.com/dotnet/roslyn/issues/161 for an auto